File: sdk/nodejs/test/test_ask.mjs
// Phase 7g.5 — tests for the natural-language → SQL surface.
//
// Run after `npm run build` produces sqlrite.<platform>-<arch>.node:
//
//     npm test
//
// (`test_ask.mjs` is picked up by the same `node --test test/`
// invocation as `test.mjs` — see package.json's test script.)
//
// Three layers covered, mirroring the Python SDK's test_ask.py:
//
// 1. **AskConfig** — option-object construction, defaults, getters,
//    fromEnv, toString deliberately omits the API key value.
// 2. **db.ask() error paths** — missing API key surfaces a clean
//    Error; closed database rejects.
// 3. **db.ask() / db.askRun() happy path** against a localhost HTTP
//    mock — the mock runs in a worker_thread to bypass Node's
//    main-event-loop deadlock that ureq's blocking sync POST would
//    otherwise create. Same shape as the Python SDK's mock; same
//    insight (napi-rs holds the JS thread for the duration of the
//    Rust call, so an in-event-loop server can't respond).

import { test } from 'node:test';
import assert from 'node:assert/strict';
import { Worker } from 'node:worker_threads';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';

import { Database, AskConfig } from '../index.js';

const __dirname = dirname(fileURLToPath(import.meta.url));

// ---------------------------------------------------------------------------
// Mock server bootstrap
//
// Spawned in a worker_thread so the main event loop is free to be
// blocked by the synchronous napi `db.ask()` call. The worker
// listens on an OS-assigned port (port 0), reports it back via
// postMessage, then serves one canned response per request and
// records what it received.
//
// Each test gets its own worker — keeps captured-request state
// per-test and avoids server-restart races between tests.

function startMockServer({ status = 200, body } = {}) {
  return new Promise((resolve, reject) => {
    const worker = new Worker(join(__dirname, 'mock-llm-server.mjs'), {
      // `body ?? SUCCESS_BODY` so a test that doesn't pass body
      // gets the standard success payload — important because the
      // worker can't see SUCCESS_BODY (different module scope).
      workerData: { status, body: body ?? SUCCESS_BODY },
    });
    worker.once('message', (msg) => {
      if (msg.type === 'ready') {
        // Wrap with a `stop()` that asks the worker to drain its
        // last captured request, then terminate.
        const stop = () =>
          new Promise((res) => {
            const onCaptured = (m) => {
              if (m.type === 'captured') {
                worker.removeListener('message', onCaptured);
                worker.terminate().then(() => res(m.captured));
              }
            };
            worker.on('message', onCaptured);
            worker.postMessage({ type: 'stop' });
          });
        resolve({ baseUrl: msg.baseUrl, stop });
      } else {
        reject(new Error(`unexpected worker message: ${JSON.stringify(msg)}`));
      }
    });
    worker.once('error', reject);
  });
}

const SUCCESS_BODY = {
  id: 'msg_test',
  type: 'message',
  role: 'assistant',
  model: 'claude-sonnet-4-6',
  content: [
    {
      type: 'text',
      text: '{"sql": "SELECT id, name FROM users", "explanation": "lists users"}',
    },
  ],
  stop_reason: 'end_turn',
  usage: {
    input_tokens: 1234,
    output_tokens: 56,
    cache_creation_input_tokens: 1000,
    cache_read_input_tokens: 0,
  },
};

// ---------------------------------------------------------------------------
// AskConfig construction + defaults

test('AskConfig defaults match the Rust side', () => {
  const cfg = new AskConfig();
  assert.equal(cfg.provider, 'anthropic');
  assert.equal(cfg.model, 'claude-sonnet-4-6');
  assert.equal(cfg.maxTokens, 1024);
  assert.equal(cfg.cacheTtl, '5m');
  assert.equal(cfg.hasApiKey, false);
});

test('AskConfig accepts an option object', () => {
  const cfg = new AskConfig({
    apiKey: 'sk-ant-test',
    model: 'claude-haiku-4-5',
    maxTokens: 2048,
    cacheTtl: '1h',
  });
  assert.equal(cfg.hasApiKey, true);
  assert.equal(cfg.model, 'claude-haiku-4-5');
  assert.equal(cfg.maxTokens, 2048);
  assert.equal(cfg.cacheTtl, '1h');
});

test('AskConfig cacheTtl accepts aliases', () => {
  for (const raw of ['5m', '5min', '5MIN', '5minutes']) {
    assert.equal(new AskConfig({ cacheTtl: raw }).cacheTtl, '5m');
  }
  for (const raw of ['1h', '1hr', '1HOUR']) {
    assert.equal(new AskConfig({ cacheTtl: raw }).cacheTtl, '1h');
  }
  for (const raw of ['off', 'none', 'DISABLED']) {
    assert.equal(new AskConfig({ cacheTtl: raw }).cacheTtl, 'off');
  }
});

test('AskConfig rejects unknown provider', () => {
  assert.throws(
    () => new AskConfig({ provider: 'openai' }),
    /unknown provider/,
  );
});

test('AskConfig rejects unknown cacheTtl', () => {
  assert.throws(
    () => new AskConfig({ cacheTtl: 'forever' }),
    /unknown cacheTtl/,
  );
});

test('AskConfig.toString does NOT leak the API key value', () => {
  const cfg = new AskConfig({ apiKey: 'sk-ant-supersecret' });
  const s = cfg.toString();
  assert.equal(s.includes('sk-ant-supersecret'), false);
  assert.equal(s.includes('<set>'), true);

  const cfg2 = new AskConfig();
  assert.equal(cfg2.toString().includes('null'), true);
});

test('AskConfig empty apiKey treated as not-set', () => {
  // Matches the Rust from_env behavior: empty string → None.
  const cfg = new AskConfig({ apiKey: '' });
  assert.equal(cfg.hasApiKey, false);
});

// ---------------------------------------------------------------------------
// AskConfig.fromEnv — env-var snapshot/restore around each test.
// node:test doesn't have an autouse fixture; we just save+restore
// inline.

function withEnvIsolation(fn) {
  const keys = [
    'SQLRITE_LLM_PROVIDER',
    'SQLRITE_LLM_API_KEY',
    'SQLRITE_LLM_MODEL',
    'SQLRITE_LLM_MAX_TOKENS',
    'SQLRITE_LLM_CACHE_TTL',
  ];
  const before = Object.fromEntries(keys.map((k) => [k, process.env[k]]));
  for (const k of keys) delete process.env[k];
  try {
    fn();
  } finally {
    for (const k of keys) {
      if (before[k] === undefined) {
        delete process.env[k];
      } else {
        process.env[k] = before[k];
      }
    }
  }
}

test('AskConfig.fromEnv with no env returns defaults + no key', () => {
  withEnvIsolation(() => {
    const cfg = AskConfig.fromEnv();
    assert.equal(cfg.provider, 'anthropic');
    assert.equal(cfg.model, 'claude-sonnet-4-6');
    assert.equal(cfg.hasApiKey, false);
    assert.equal(cfg.cacheTtl, '5m');
  });
});

test('AskConfig.fromEnv reads SQLRITE_LLM_* vars', () => {
  withEnvIsolation(() => {
    process.env.SQLRITE_LLM_API_KEY = 'env-key';
    process.env.SQLRITE_LLM_MODEL = 'claude-haiku-4-5';
    process.env.SQLRITE_LLM_MAX_TOKENS = '512';
    process.env.SQLRITE_LLM_CACHE_TTL = '1h';

    const cfg = AskConfig.fromEnv();
    assert.equal(cfg.hasApiKey, true);
    assert.equal(cfg.model, 'claude-haiku-4-5');
    assert.equal(cfg.maxTokens, 512);
    assert.equal(cfg.cacheTtl, '1h');
  });
});

test('AskConfig.fromEnv rejects invalid SQLRITE_LLM_MAX_TOKENS', () => {
  withEnvIsolation(() => {
    process.env.SQLRITE_LLM_MAX_TOKENS = 'not-an-int';
    assert.throws(() => AskConfig.fromEnv(), /MAX_TOKENS/);
  });
});

// ---------------------------------------------------------------------------
// db.ask() error paths (no LLM call needed)

test('db.ask without API key surfaces a clean error', () => {
  withEnvIsolation(() => {
    const db = new Database(':memory:');
    try {
      assert.throws(() => db.ask('How many users?'), /missing API key/);
    } finally {
      db.close();
    }
  });
});

test('db.ask on a closed database rejects', () => {
  const db = new Database(':memory:');
  db.close();
  assert.throws(() => db.ask('anything'), /closed/);
});

test('setAskConfig with no key then ask still raises', () => {
  withEnvIsolation(() => {
    const db = new Database(':memory:');
    try {
      const cfg = new AskConfig(); // no apiKey
      db.setAskConfig(cfg);
      assert.throws(() => db.ask('anything'), /missing API key/);
    } finally {
      db.close();
    }
  });
});

// ---------------------------------------------------------------------------
// db.ask() happy path against the worker-thread localhost mock

test('db.ask returns parsed AskResponse', async () => {
  const db = new Database(':memory:');
  try {
    db.exec('CREATE TABLE users (id INTEGER PRIMARY KEY, name TEXT, age INTEGER)');
    const mock = await startMockServer();
    let resp;
    try {
      const cfg = new AskConfig({ apiKey: 'test-key', baseUrl: mock.baseUrl });
      resp = db.ask('How many users are over 30?', cfg);
    } finally {
      await mock.stop();
    }
    assert.equal(resp.sql, 'SELECT id, name FROM users');
    assert.equal(resp.explanation, 'lists users');
    assert.equal(resp.usage.inputTokens, 1234);
    assert.equal(resp.usage.cacheCreationInputTokens, 1000);
    assert.equal(resp.usage.cacheReadInputTokens, 0);
  } finally {
    db.close();
  }
});

test('db.ask sends schema + cache-control + auth headers', async () => {
  const db = new Database(':memory:');
  try {
    db.exec('CREATE TABLE users (id INTEGER PRIMARY KEY, name TEXT)');
    const mock = await startMockServer();
    try {
      const cfg = new AskConfig({ apiKey: 'test-key', baseUrl: mock.baseUrl });
      db.ask('list users', cfg);
    } finally {
      var captured = await mock.stop();
    }
    assert.equal(captured.body.model, 'claude-sonnet-4-6');
    assert.equal(captured.body.max_tokens, 1024);
    assert.equal(captured.body.messages[0].role, 'user');
    assert.equal(captured.body.messages[0].content, 'list users');
    assert.match(captured.body.system[1].text, /CREATE TABLE users/);
    assert.equal(captured.body.system[1].cache_control.type, 'ephemeral');
    // Headers (lowercased by the worker for case-insensitive lookup).
    assert.equal(captured.headers['x-api-key'], 'test-key');
    assert.equal(captured.headers['anthropic-version'], '2023-06-01');
  } finally {
    db.close();
  }
});

test('setAskConfig persists across calls', async () => {
  const db = new Database(':memory:');
  try {
    db.exec('CREATE TABLE t (id INTEGER PRIMARY KEY)');
    const mock = await startMockServer();
    let r1, r2;
    try {
      const cfg = new AskConfig({ apiKey: 'persisted', baseUrl: mock.baseUrl });
      db.setAskConfig(cfg);
      r1 = db.ask('first');
      r2 = db.ask('second');
    } finally {
      await mock.stop();
    }
    assert.equal(r1.sql, 'SELECT id, name FROM users');
    assert.equal(r2.sql, 'SELECT id, name FROM users');
  } finally {
    db.close();
  }
});

test('db.askRun executes the generated SQL and returns rows', async () => {
  const db = new Database(':memory:');
  try {
    db.exec('CREATE TABLE users (id INTEGER PRIMARY KEY, name TEXT, age INTEGER)');
    db.exec("INSERT INTO users (name, age) VALUES ('alice', 30)");
    db.exec("INSERT INTO users (name, age) VALUES ('bob', 25)");
    const mock = await startMockServer();
    let rows;
    try {
      const cfg = new AskConfig({ apiKey: 'test-key', baseUrl: mock.baseUrl });
      rows = db.askRun('list users', cfg);
    } finally {
      await mock.stop();
    }
    assert.equal(rows.length, 2);
    const names = rows.map((r) => r.name).sort();
    assert.deepEqual(names, ['alice', 'bob']);
  } finally {
    db.close();
  }
});

test('db.askRun on empty SQL response throws with model explanation', async () => {
  const declineBody = { ...SUCCESS_BODY };
  declineBody.content = [
    {
      type: 'text',
      text: '{"sql": "", "explanation": "schema lacks a widgets table"}',
    },
  ];
  const db = new Database(':memory:');
  try {
    const mock = await startMockServer({ body: declineBody });
    try {
      const cfg = new AskConfig({ apiKey: 'test-key', baseUrl: mock.baseUrl });
      assert.throws(
        () => db.askRun('how many widgets?', cfg),
        /declined.*widgets table/,
      );
    } finally {
      await mock.stop();
    }
  } finally {
    db.close();
  }
});

test('API 4xx response surfaces as JS Error with the structured message', async () => {
  const db = new Database(':memory:');
  try {
    const mock = await startMockServer({
      status: 400,
      body: {
        type: 'error',
        error: {
          type: 'invalid_request_error',
          message: 'max_tokens too large',
        },
      },
    });
    try {
      const cfg = new AskConfig({ apiKey: 'test-key', baseUrl: mock.baseUrl });
      assert.throws(
        () => db.ask('anything', cfg),
        /400.*invalid_request_error.*max_tokens too large/,
      );
    } finally {
      await mock.stop();
    }
  } finally {
    db.close();
  }
});

File: sdk/python/tests/test_ask.py
"""Tests for the Phase 7g.4 natural-language → SQL surface.

Three layers covered:

1. **AskConfig** — construction (kwargs + from_env), defaults,
   attribute access, env-var precedence.
2. **conn.ask() error paths** — missing-API-key surfaces a clear
   `SQLRiteError`; closed connection rejects.
3. **conn.ask() happy path** against a localhost HTTP mock — full
   round-trip through PyO3 → engine → sqlrite-ask → ureq POST →
   our mock server → AskResponse parsing → Python types. The mock
   stands in for api.anthropic.com so CI doesn't need real
   credentials and we don't pay per test run.

The mock uses Python's stdlib `http.server` — no extra deps.

Run after `maturin develop` (or against an installed wheel):

    python -m pytest sdk/python/tests/test_ask.py
"""

from __future__ import annotations

import json
import os
import threading
from http.server import BaseHTTPRequestHandler, HTTPServer
from typing import Any

import pytest

import sqlrite


# ---------------------------------------------------------------------------
# AskConfig construction + defaults


class TestAskConfigConstruction:
    def test_default_kwargs(self):
        cfg = sqlrite.AskConfig()
        assert cfg.provider == "anthropic"
        assert cfg.model == "claude-sonnet-4-6"
        assert cfg.max_tokens == 1024
        assert cfg.cache_ttl == "5m"
        assert cfg.api_key is None

    def test_explicit_kwargs(self):
        cfg = sqlrite.AskConfig(
            api_key="sk-ant-test",
            model="claude-haiku-4-5",
            max_tokens=2048,
            cache_ttl="1h",
        )
        assert cfg.api_key == "sk-ant-test"
        assert cfg.model == "claude-haiku-4-5"
        assert cfg.max_tokens == 2048
        assert cfg.cache_ttl == "1h"

    def test_cache_ttl_aliases(self):
        # Each alias the Rust side accepts should round-trip to the
        # canonical short form.
        for raw in ("5m", "5min", "5minutes", "5M"):
            cfg = sqlrite.AskConfig(cache_ttl=raw)
            assert cfg.cache_ttl == "5m", f"got {cfg.cache_ttl!r} for input {raw!r}"
        for raw in ("1h", "1hr", "1hour", "1H"):
            cfg = sqlrite.AskConfig(cache_ttl=raw)
            assert cfg.cache_ttl == "1h"
        for raw in ("off", "none", "disabled", "OFF"):
            cfg = sqlrite.AskConfig(cache_ttl=raw)
            assert cfg.cache_ttl == "off"

    def test_unknown_provider_raises(self):
        with pytest.raises(sqlrite.SQLRiteError, match="unknown provider"):
            sqlrite.AskConfig(provider="openai")  # not yet supported

    def test_unknown_cache_ttl_raises(self):
        with pytest.raises(sqlrite.SQLRiteError, match="unknown cache_ttl"):
            sqlrite.AskConfig(cache_ttl="forever")

    def test_repr_does_not_leak_api_key(self):
        cfg = sqlrite.AskConfig(api_key="sk-ant-supersecret")
        # Whether the key is present or not should be visible in repr,
        # but the actual key value MUST NOT appear (so accidentally
        # printing config in logs doesn't leak).
        r = repr(cfg)
        assert "sk-ant-supersecret" not in r
        assert "<set>" in r

        cfg2 = sqlrite.AskConfig()
        r2 = repr(cfg2)
        assert "None" in r2

    def test_empty_api_key_treated_as_none(self):
        # Matches the Rust from_env behavior: empty string → None.
        cfg = sqlrite.AskConfig(api_key="")
        assert cfg.api_key is None


class TestAskConfigFromEnv:
    """Verify env-var defaults match the Rust side. Each test
    snapshots the current env, mutates, and restores."""

    @pytest.fixture(autouse=True)
    def env_isolation(self):
        keys = (
            "SQLRITE_LLM_PROVIDER",
            "SQLRITE_LLM_API_KEY",
            "SQLRITE_LLM_MODEL",
            "SQLRITE_LLM_MAX_TOKENS",
            "SQLRITE_LLM_CACHE_TTL",
        )
        before = {k: os.environ.get(k) for k in keys}
        # Clear all first so tests start from a known state.
        for k in keys:
            os.environ.pop(k, None)
        yield
        for k, v in before.items():
            if v is None:
                os.environ.pop(k, None)
            else:
                os.environ[k] = v

    def test_no_env_returns_defaults_with_none_key(self):
        cfg = sqlrite.AskConfig.from_env()
        assert cfg.provider == "anthropic"
        assert cfg.model == "claude-sonnet-4-6"
        assert cfg.api_key is None
        assert cfg.cache_ttl == "5m"

    def test_env_overrides_defaults(self):
        os.environ["SQLRITE_LLM_API_KEY"] = "env-key"
        os.environ["SQLRITE_LLM_MODEL"] = "claude-haiku-4-5"
        os.environ["SQLRITE_LLM_MAX_TOKENS"] = "512"
        os.environ["SQLRITE_LLM_CACHE_TTL"] = "1h"

        cfg = sqlrite.AskConfig.from_env()
        assert cfg.api_key == "env-key"
        assert cfg.model == "claude-haiku-4-5"
        assert cfg.max_tokens == 512
        assert cfg.cache_ttl == "1h"

    def test_invalid_max_tokens_raises(self):
        os.environ["SQLRITE_LLM_MAX_TOKENS"] = "not-an-int"
        with pytest.raises(sqlrite.SQLRiteError, match="MAX_TOKENS"):
            sqlrite.AskConfig.from_env()


# ---------------------------------------------------------------------------
# conn.ask() error paths


class TestAskErrorPaths:
    @pytest.fixture(autouse=True)
    def env_isolation(self):
        before = os.environ.pop("SQLRITE_LLM_API_KEY", None)
        yield
        if before is not None:
            os.environ["SQLRITE_LLM_API_KEY"] = before

    def test_missing_api_key_raises_clear_error(self):
        conn = sqlrite.connect(":memory:")
        try:
            with pytest.raises(sqlrite.SQLRiteError, match="missing API key"):
                conn.ask("How many users?")
        finally:
            conn.close()

    def test_closed_connection_rejects(self):
        conn = sqlrite.connect(":memory:")
        conn.close()
        with pytest.raises(sqlrite.SQLRiteError, match="closed"):
            conn.ask("anything")

    def test_set_ask_config_with_no_key_then_ask_still_raises(self):
        # Setting a config without an api_key isn't auto-promoted to
        # an env lookup — explicit None means "use this exact config",
        # which has no key, which fails.
        conn = sqlrite.connect(":memory:")
        try:
            cfg = sqlrite.AskConfig()  # api_key=None
            conn.set_ask_config(cfg)
            with pytest.raises(sqlrite.SQLRiteError, match="missing API key"):
                conn.ask("anything")
        finally:
            conn.close()


# ---------------------------------------------------------------------------
# conn.ask() happy path against a localhost HTTP mock


_MOCK_RESPONSE = {
    "id": "msg_test",
    "type": "message",
    "role": "assistant",
    "model": "claude-sonnet-4-6",
    "content": [
        {
            "type": "text",
            # Inner JSON quotes are escaped — this is what the model
            # would emit per our prompt template.
            "text": '{"sql": "SELECT COUNT(*) FROM users", "explanation": "counts users"}',
        }
    ],
    "stop_reason": "end_turn",
    "usage": {
        "input_tokens": 1234,
        "output_tokens": 56,
        "cache_creation_input_tokens": 1000,
        "cache_read_input_tokens": 0,
    },
}


class _MockHandler(BaseHTTPRequestHandler):
    """Captures one POST + serves a canned response. Each instance
    of MockServer below installs its own captured-request slot via a
    class attribute on a freshly-subclassed handler; it's not pretty
    but it avoids globals."""

    captured: dict[str, Any] = {}
    response_status: int = 200
    response_body: dict[str, Any] = _MOCK_RESPONSE

    def do_POST(self) -> None:  # noqa: N802 — http.server naming
        length = int(self.headers.get("Content-Length", "0"))
        body = self.rfile.read(length).decode("utf-8") if length else ""
        type(self).captured = {
            "path": self.path,
            "headers": {k.lower(): v for k, v in self.headers.items()},
            "body": json.loads(body) if body else None,
        }
        self.send_response(type(self).response_status)
        self.send_header("Content-Type", "application/json")
        self.end_headers()
        self.wfile.write(json.dumps(type(self).response_body).encode("utf-8"))

    def log_message(self, *args, **kwargs) -> None:  # silence stderr noise
        pass


class _MockServer:
    """Tiny localhost HTTP server. Use as a context manager; yields
    the base URL to point AskConfig at."""

    def __init__(
        self,
        status: int = 200,
        body: dict[str, Any] | None = None,
    ) -> None:
        # Subclass the handler so per-test status / body customizations
        # don't bleed across tests.
        cls_name = f"_Handler_{id(self)}"
        attrs = {
            "captured": {},
            "response_status": status,
            "response_body": body if body is not None else _MOCK_RESPONSE,
        }
        self.handler_cls = type(cls_name, (_MockHandler,), attrs)
        self.server = HTTPServer(("127.0.0.1", 0), self.handler_cls)
        self.thread = threading.Thread(target=self.server.serve_forever, daemon=True)

    def __enter__(self) -> str:
        self.thread.start()
        host, port = self.server.server_address
        return f"http://{host}:{port}"

    def __exit__(self, *exc) -> None:
        self.server.shutdown()
        self.server.server_close()
        self.thread.join(timeout=2)

    @property
    def captured(self) -> dict[str, Any]:
        return self.handler_cls.captured  # type: ignore[no-any-return]


class TestAskHappyPath:
    @pytest.fixture(autouse=True)
    def env_isolation(self):
        before = {
            k: os.environ.pop(k, None)
            for k in (
                "SQLRITE_LLM_API_KEY",
                "SQLRITE_LLM_MODEL",
                "SQLRITE_LLM_BASE_URL",
            )
        }
        yield
        for k, v in before.items():
            if v is not None:
                os.environ[k] = v

    def test_ask_returns_parsed_response(self):
        conn = sqlrite.connect(":memory:")
        try:
            conn.execute(
                "CREATE TABLE users (id INTEGER PRIMARY KEY, name TEXT, age INTEGER)"
            )
            with _MockServer() as base_url:
                cfg = sqlrite.AskConfig(api_key="test-key", base_url=base_url)
                resp = conn.ask("How many users are over 30?", cfg)
            assert resp.sql == "SELECT COUNT(*) FROM users"
            assert resp.explanation == "counts users"
            assert resp.usage.input_tokens == 1234
            assert resp.usage.cache_creation_input_tokens == 1000
            assert resp.usage.cache_read_input_tokens == 0
        finally:
            conn.close()

    def test_ask_request_body_shape(self):
        """Verify the request payload matches what the Anthropic API
        expects: model, max_tokens, system blocks (with cache_control
        on the schema), messages, and the right headers."""
        conn = sqlrite.connect(":memory:")
        try:
            conn.execute("CREATE TABLE users (id INTEGER PRIMARY KEY, name TEXT)")
            server = _MockServer()
            base_url = server.__enter__()
            try:
                cfg = sqlrite.AskConfig(api_key="test-key", base_url=base_url)
                conn.ask("list users", cfg)
                # Inspect what the server received
                body = server.captured["body"]
                assert body["model"] == "claude-sonnet-4-6"
                assert body["max_tokens"] == 1024
                assert body["messages"][0]["role"] == "user"
                assert body["messages"][0]["content"] == "list users"
                # Schema block carries the CREATE TABLE.
                assert "CREATE TABLE users" in body["system"][1]["text"]
                # Cache marker on the schema block (default 5m TTL).
                assert body["system"][1]["cache_control"]["type"] == "ephemeral"
                # Headers: x-api-key + anthropic-version.
                hdrs = server.captured["headers"]
                assert hdrs.get("x-api-key") == "test-key"
                assert hdrs.get("anthropic-version") == "2023-06-01"
            finally:
                server.__exit__(None, None, None)
        finally:
            conn.close()

    def test_set_ask_config_persists_across_calls(self):
        conn = sqlrite.connect(":memory:")
        try:
            conn.execute("CREATE TABLE t (id INTEGER PRIMARY KEY)")
            with _MockServer() as base_url:
                cfg = sqlrite.AskConfig(api_key="persisted", base_url=base_url)
                conn.set_ask_config(cfg)
                # No config arg on subsequent calls — should reuse.
                resp1 = conn.ask("first")
                resp2 = conn.ask("second")
                assert resp1.sql == resp2.sql == "SELECT COUNT(*) FROM users"
        finally:
            conn.close()

    def test_ask_run_executes_generated_sql(self):
        # Use a mock that returns a SELECT shape the engine actually
        # supports. (No COUNT(*) — aggregates are deferred to a future
        # phase.) The point of this test isn't what SQL the LLM
        # generates — it's that ask_run() actually executes whatever
        # comes back through cursor.execute().
        executable_body = dict(_MOCK_RESPONSE)
        executable_body["content"] = [
            {
                "type": "text",
                "text": '{"sql": "SELECT id, name FROM users", "explanation": "lists users"}',
            }
        ]
        conn = sqlrite.connect(":memory:")
        try:
            conn.execute(
                "CREATE TABLE users (id INTEGER PRIMARY KEY, name TEXT, age INTEGER)"
            )
            conn.execute("INSERT INTO users (name, age) VALUES ('alice', 30)")
            conn.execute("INSERT INTO users (name, age) VALUES ('bob', 25)")
            with _MockServer(body=executable_body) as base_url:
                cfg = sqlrite.AskConfig(api_key="test-key", base_url=base_url)
                cur = conn.ask_run("list users", cfg)
                rows = cur.fetchall()
                assert len(rows) == 2
                assert {row[1] for row in rows} == {"alice", "bob"}
        finally:
            conn.close()

    def test_per_call_config_overrides_per_connection(self):
        conn = sqlrite.connect(":memory:")
        try:
            conn.execute("CREATE TABLE t (id INTEGER PRIMARY KEY)")
            with _MockServer() as base_url:
                # Per-connection: bad key (wouldn't work)
                conn.set_ask_config(
                    sqlrite.AskConfig(api_key="stale-key", base_url=base_url)
                )
                # Per-call: correct key. Should win.
                per_call = sqlrite.AskConfig(api_key="correct-key", base_url=base_url)
                conn.ask("anything", per_call)
            # If we got here without an exception, per-call config won
            # (stale-key would have… well, both work against the mock,
            # but the assertion is that there's no error). The header
            # assertion in test_ask_request_body_shape covers the
            # actual key sent.
        finally:
            conn.close()


class TestAskApiErrorSurfacing:
    def test_400_response_surfaces_as_sqlrite_error(self):
        conn = sqlrite.connect(":memory:")
        try:
            error_body = {
                "type": "error",
                "error": {
                    "type": "invalid_request_error",
                    "message": "max_tokens too large",
                },
            }
            with _MockServer(status=400, body=error_body) as base_url:
                cfg = sqlrite.AskConfig(api_key="test-key", base_url=base_url)
                with pytest.raises(sqlrite.SQLRiteError) as exc_info:
                    conn.ask("anything", cfg)
                msg = str(exc_info.value)
                assert "400" in msg
                assert "invalid_request_error" in msg
                assert "max_tokens too large" in msg
        finally:
            conn.close()

    def test_ask_run_on_empty_sql_response_raises(self):
        """Model declined to generate SQL — ask_run() must raise
        rather than execute the empty string (which would surface as
        a less-helpful parser error)."""
        conn = sqlrite.connect(":memory:")
        try:
            decline_body = dict(_MOCK_RESPONSE)
            decline_body["content"] = [
                {
                    "type": "text",
                    "text": (
                        '{"sql": "", "explanation": '
                        '"the schema does not contain a users table"}'
                    ),
                }
            ]
            with _MockServer(body=decline_body) as base_url:
                cfg = sqlrite.AskConfig(api_key="test-key", base_url=base_url)
                with pytest.raises(sqlrite.SQLRiteError, match="declined"):
                    conn.ask_run("how many widgets?", cfg)
        finally:
            conn.close()
